pkg/registry/parse.go

```diff
@@ -0,0 +1,253 @@
+package registry
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+type bundleParser struct {
+	log *logrus.Entry
+}
+
+func newBundleParser(log *logrus.Entry) *bundleParser {
+	return &bundleParser{
+		log: log,
+	}
+}
+
+// Parse parses the given FS into a Bundle.
+func (b *bundleParser) Parse(root fs.FS) (*Bundle, error) {
+	if root == nil {
+		return nil, fmt.Errorf("filesystem is nil")
+	}
+
+	bundle := &Bundle{}
+	manifests, err := fs.Sub(root, "manifests")
+	if err != nil {
+		return nil, fmt.Errorf("error opening manifests directory: %s", err)
+	}
+	if err := b.addManifests(manifests, bundle); err != nil {
+		return nil, err
+	}
+
+	metadata, err := fs.Sub(root, "metadata")
+	if err != nil {
+		return nil, fmt.Errorf("error opening metadata directory: %s", err)
+	}
+	if err := b.addMetadata(metadata, bundle); err != nil {
+		return nil, err
+	}
+
+	derived, err := b.derivedProperties(bundle)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive properties: %s", err)
+	}
+
+	bundle.Properties = propertySet(append(bundle.Properties, derived...))
+
+	return bundle, nil
+}
+
+// addManifests adds the result of parsing the manifests directory to a bundle.
+func (b *bundleParser) addManifests(manifests fs.FS, bundle *Bundle) error {
+	files, err := fs.ReadDir(manifests, ".")
+	if err != nil {
+		return err
+	}
+
+	var csvFound bool
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+
+		name := f.Name()
+		if strings.HasPrefix(name, ".") {
+			continue
+		}
+
+		obj := &unstructured.Unstructured{}
+		if err = decodeFileFS(manifests, name, obj); err != nil {
+			b.log.Warnf("failed to decode: %s", err)
+			continue
+		}
+
+		// Only include the first CSV we find in the
```

Comment:
is this behavior documented anywhere (that we only care about the first CSV in the bundle that we find?) I'm not sure how `ReadDir` lists files, if it's alphabetical/lexicographic or random. Can users with more than 1 CSV in their bundle know how this behaves deterministically? 

I thought this would be an error condition, if anything other than exactly 1 CSV is present in the bundle.

Reply:
> I'm not sure how ReadDir lists files, if it's alphabetical/lexicographic or random.

Very good points!

I don't think we can assume anything about the order files are returned by implementations of `fs.FS`.

I know that 1 CSV per bundle is an invariant for bundles, although I'm unsure where -- if at all -- this is documented.

I think it behooves us to have some deterministic behavior, but I don't want to stoop to some arbitrary heuristic to do so.

I'll dig up what the behavior is for the old parse logic and see if we want to replicate that (it might return an error)

Reply:
IIRC the first CSV found by `filepath.Walk()` was used, and finding another CSV afterwards resulted in an error.

Reply:
After reviewing the code again, it _does_ seem like the behavior -- pick the first CSV and ignore the rest -- matches what's currently in `master`.

See https://github.com/operator-framework/operator-registry/blob/f1629cd112099303272c245f1f9de3f8ab6fdcf2/pkg/registry/imageinput.go#L90 for more details.